source/common/http/conn_manager_impl.cc

```diff
@@ -1074,6 +1077,26 @@ void ConnectionManagerImpl::ActiveStreamDecoderFilter::
   parent_.connection_manager_.stats_.named_.downstream_flow_control_resumed_reading_total_.inc();
 }
 
+void ConnectionManagerImpl::ActiveStreamDecoderFilter::addDownstreamWatermarkCallbacks(
+    DownstreamWatermarkCallbacks& watermark_callbacks) {
+  // This is called exactly once per stream, by the router filter.
+  // If there's ever a need for another filter to subscribe to watermark callbacks this can be
+  // turned into a vector.
+  ASSERT(parent_.watermark_callbacks_ == nullptr);
+  parent_.watermark_callbacks_ = &watermark_callbacks;
+  for (uint32_t i = 0; i < parent_.high_watermark_count_; ++i) {
```

Comment:
I think this is right. 

Imagine a H2 connection where the Network::Connection gets backed up.  in callHighWatermarkCallbacks you inform all active streams, and track that the overall unit of the H2 connection is read-disabled.

Now you create a new stream and associate it with a new upstream connection, you need to get it in sync with all the other streams by calling readDisable the correct number of times.   If it's not blocked, you don't need to do anything.

Meanwhile if that connection wraps up and the upstream connection is disassociated in a way that can be reused (HTTP), you need to unravel the readDisables somewhere (I chose here for consistency but we could do it where the connection is disassociated or re-associated with another request.